### PR TITLE
Fix Daylight saving transition handling

### PR DESCRIPTION
**Pull request by notnoop**
This fixes handling of daylight saving transitions.  It does so in the following way:

* If the cronexpr isn't satisfied on a day due to a leap forward, the day will be skipped from the periodic job.
  * e.g. a periodic job specifying `30 2 * * *` will not run on 2019-03-10 in the US/New_York timezone

* If an hour is "repeated" due to daylight savings backward transition, the job may run twice that day
  * `30 1 * * *` will run twice on 2019-11-03 US/New_York - once at 1:30am EDT and then again at 1:30am EST.

If a job sets a cron expression that runs every second/minute/hour, the job will run as expected during the entire time and not miss any period.

Prior to the change, the job may run at somewhat arbitrary time and may cause a job to be running uncontrollably and causing a denial of service.

This approach is relatively simple easy to reason approach.  Operators are **strongly** encouraged to use UTC, or have their nightly jobs run in times that aren't affected by day light saving changes (e.g. 12.30am or 3.30am ET) .

### Implementation Notes

I have forked github.com/gorhill/cronexpr in https://github.com/hashicorp/cronexpr .  I have updated it so that it is daylight saving aware.  The relevant PR is https://github.com/hashicorp/cronexpr/pull/1 .  It includes the tests here and more, specifically for US (representing the normal case where DST is hour offset) and Lord Howe Island, Australia (where daylight saving has a 30 minute offset).

Also, this also adds checking if a cron expression is valid.  We'd better report if a cron expression is invalid.  Previously, we'd ignore the periodic job silently.

### Abandoned ideas

I have considered an alternative approach that's closer to how some cron distributions manage daylight saving changes.  Namely, running skipped jobs immediately when time leaps forward.  Also, considered ensuring that jobs that are expected to run once a day run only once backward transition period.

However, doing so will complicate our periodic job handling - specially considering it's a stateless processor now.  Also, deciding whether a job is expected to run only once during the backward transition period isn't trivial and will probably may surprise uses just as much (e.g. how should `* */2 * * *` be handled? ).

Fixes https://github.com/hashicorp/nomad/issues/7289
Fixes https://github.com/hashicorp/nomad/issues/5410

### CODE REVIEW

File: api/go.mod [hunk 1]
```diff
@@ -5,8 +5,8 @@ go 1.12
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/docker/go-units v0.3.3
-	github.com/gorhill/cronexpr v0.0.0-20180427100037-88b0669f7d75
 	github.com/gorilla/websocket v1.4.1
+	github.com/hashicorp/cronexpr v0.0.0-20200507212857-921335d977b6
```
Comment (shoenig) [inline api/go.mod:9]: Since we own the library now, can we semver it?

**Comment by shoenig** [comment]
Should we add a note on the [time_zone](https://www.nomadproject.io/docs/job-specification/periodic/#time_zone) parameter about skipping jobs for times that do not exist?

**Comment by notnoop** [comment]
> Should we add a note on the time_zone parameter about skipping jobs for times that do not exist?

 
Yeah, I can update the documentation and upgrade guide in a follow up after getting 👍 on this approach.

**Comment by shoenig** [comment]
FYI one of the conditions of the [APLv2 ](https://choosealicense.com/licenses/apache-2.0/)from cronexpr is that we explicitly document the changes made. Since we forked it, we should probably update the changed files with a notice. 
```
You must cause any modified files to carry prominent notices
          stating that You changed the files; and [...]
```

**Comment by notnoop** [comment]
Thanks @schmichael .  I'll merge this to be in 0.11.2 - and follow up with docs in a separate PR.

**Pull request merged**